### ISSUE TITLE
How does MLV App deal with dropped frames?

### ISSUE THREAD

**Issue by ilia3101** [ilia3101/MLV-App#183]
Especially if there's audio

**Comment by masc4ii** [comment]
Hm... I never had a file with a dropped frame, so I can't tell for sure. I would expect at least on export, that the audio would loose sync on a dropped frame... but I never tried.

**Comment by ilia3101** [comment]
Yeah I guess most people have drop frames disabled, but you can enable it. I will have to investigate. I wonder if the camera numbers the frames with a gap to indicate that there is one missing.

**Comment by bouncyball-git** [comment]
@ilia3101: I've implemented hardware timestamp sorting long time ago ;) (frame_index_sort() function in video_mlv.c)

Masc supplied me with the awful MLV with unsorted frames. It played with chops and hiccups before I changed MLV reading alg. Now it handles this file almost perfectly (almost because this MLV has dropped frames too)

@masc4ii: do you still have that MLV - some machine moving it's arm (like printer) :)

**Comment by masc4ii** [comment]
I am not sure if I still have it. Problem: I don't see those problems 😄 
I know that you're sorting those frames using the timestamp. But what happens when a frame is missing? E.g. do we show the previous frame twice?

**Comment by bouncyball-git** [comment]
Also I put the brute force code (video_mlv.c) to handle corrupted header MLVs. I have one sample MLV (someone posted on the forum, girl walks on the street and some guy comes along with her and talks to her...)  and it is working nicely. Loading a bit slower but finally indexing is done and file plays fine.

**Comment by bouncyball-git** [comment]
> I am not sure if I still have it. Problem: I don't see those problems smile
hehehe

> I know that you're sorting those frames using the timestamp. But what happens when a frame is missing? E.g. do we show the previous frame twice?
nope :( it just omits it and goes on... audio can be not synced. That MLV of yours had no audio...

**Comment by ilia3101** [comment]
> It played with chops and hiccups before I changed file reading alg simply not to rely on frame number taken from the VIDF header struct. Now it handles this file almost perfectly (almost because this MLV had dropped frames too)

So the hardware timestamp is also useful. Interesting. What is the unit of time that it counts in?

**Comment by bouncyball-git** [comment]
Microseconds.

**Comment by ilia3101** [comment]
> btw in dng.c (mlvapp) we have 2 functions: raw data unpacker to 16bit word (written by dmilligan) and 16bit to any bit packer (written by me. I'm using this to produce any bit depth DNGs). You can make use of them, your packers are may be a bit faster but anyway one function instead of 3 and the possibility to pack to ANY bit depth is not bad I think.

Wow thank you! Packing to any bitdepth is nice (I guess that means any within 8-16). I will put your function to use. But was dmilligan's unpacking code originally released with a permissive (non GPL) license? I just wanna be careful to avoid putting actual GPL stuff in the new library without original author's permission. (and what do you think of this issue in general)

**Comment by bouncyball-git** [comment]
> Wow thank you! Packing to any bitdepth is nice (I guess that means any within 8-16). I will put your function to use. 

You can go all the way down :) (step is 1bit)

> But was dmilligan's unpacking code originally released with a permissive (non GPL) license?

It is a part of MLVFS. So...

**Comment by ilia3101** [comment]
> You can go all the way down :) (step is 1bit)

Wow I can't wait to try out 1 bit MLV!
